ocp_resources/utils.py

```diff
@@ -26,74 +27,154 @@ class TimeoutSampler:
+        self.exceptions_dict = self._pre_process_exceptions(
+            exceptions=exceptions,
+            exceptions_msg=exceptions_msg,
+            exceptions_dict=exceptions_dict,
+        )
+        self._exceptions = tuple(self.exceptions_dict.keys())
+
+    def _pre_process_exceptions(self, exceptions, exceptions_msg, exceptions_dict):
+        """
+        Process exception input for use within _process_exeception()
+
+        TODO: Deprecation: This method should be removed when the 'exceptions' argument is removed from __init__
+
+        Validate exception inputs:
+            exceptions
+            exceptions_msg
+            exceptions_dict
+        """
+        output = {}
+        if exceptions_dict and (exceptions or exceptions_msg):
+            raise ValueError(
+                "Must specify either exceptions_dict or exceptions/exception_msg, not both"
+            )
+
+        elif exceptions or exceptions_msg:
+            warn(
+                "TimeoutSampler() exception and exception_msg are now deprecated. "
+                "Please update to use exceptions_dict by 2021-MM-DD",
+                DeprecationWarning,
+            )
+
+        if exceptions_dict:
+            output.update(exceptions_dict)
+
+        elif exceptions is None:
+            output[Exception] = []
+
+        else:
+            if not isinstance(exceptions, tuple):
+                exceptions = (exceptions,)
+
+            for exp in exceptions:
+                if exceptions_msg:
+                    output[exp] = [exceptions_msg]
+                else:
+                    output[exp] = []
+
+        return output
+
     @property
     def _func_log(self):
         return f"Function: {self.func} Kwargs: {self.func_kwargs}"
 
     def __iter__(self):
-        last_exp = None
         timeout_watch = TimeoutWatch(timeout=self.wait_timeout)
         if self.print_log:
             LOGGER.info(
                 f"Waiting for {self.wait_timeout} seconds, retry every {self.sleep} seconds"
             )
 
+        last_exp_log = ""
         while timeout_watch.remaining_time() > 0:
             try:
                 self.elapsed_time = self.wait_timeout - timeout_watch.remaining_time()
                 yield self.func(**self.func_kwargs)
                 self.elapsed_time = None
+
                 time.sleep(self.sleep)
 
-            except self.exception as exp:
-                last_exp = exp
+            except self._exceptions as exp:
+                last_exp_log = self._process_exeception(exp=exp)
                 self.elapsed_time = None
-
                 time.sleep(self.sleep)
 
             finally:
                 if self.elapsed_time and self.print_log:
                     LOGGER.info(f"Elapsed time: {self.elapsed_time}")
 
-        raise TimeoutExpiredError(self._process_execution(exp=last_exp))
+        raise TimeoutExpiredError(last_exp_log)
```

Comment:
```suggestion
        raise TimeoutExpiredError(_process_exeception(exp=exp))

```

Reply:
Does this apply in the new configuration?

Reply:
yes

Reply:
how? the exception has been processed.